File: native/android/PropertyCross/src/com/propertycross/android/model/PropertyDataSource.java
package com.propertycross.android.model;

import org.json.JSONException;
import org.json.JSONObject;

import com.propertycross.android.events.Callback;

public class PropertyDataSource {

    private final static String TAG = "PROPERTYDATASOURCE";
	private IJsonPropertySearch jsonPropertySearch;

	public PropertyDataSource(IJsonPropertySearch jsonPropertySearch) {
		this.jsonPropertySearch = jsonPropertySearch;
	}

	public void findProperties(String searchText, int pageNumber,
			final Callback<PropertyDataSourceResult> complete, final Callback<Exception> error) {

		jsonPropertySearch.findProperties(searchText, pageNumber,
				new Callback<String>() {
					public void complete(String result) {
						handleResponse(result, complete, error);
					}
				}, error);
	}

	public void findProperties(double latitude, double longitude, int pageNumber,
			final Callback<PropertyDataSourceResult> complete, final Callback<Exception> error) {

		jsonPropertySearch.findProperties(latitude, longitude, pageNumber,
				new Callback<String>() {
					public void complete(String result) {
						handleResponse(result, complete, error);
					}
				}, error);
	}

	private void handleResponse(String jsonResponse,
	        Callback<PropertyDataSourceResult> complete, Callback<Exception> error) {
		try {
			JSONObject json = new JSONObject(jsonResponse);
			JSONObject response = json.getJSONObject("response");
			String responseCode = response.getString("application_response_code");
			
			if (isLocationUnambiguous(responseCode) ||
			    isBestGuessLocation(responseCode) ||
			    isLargeLocation(responseCode)) {
				PropertyDataSourceResult result = new PropertyListingsResult(json);
				complete.complete(result);
			}
			else if (isLocationAmbiguous(responseCode) ||
			         isLocationMisspelled(responseCode)) {
				PropertyDataSourceResult result = new PropertyLocationsResult(json);
				complete.complete(result);
			}
			else {
				fail(complete);
			}
		}
		catch (JSONException e) {
		    android.util.Log.e(TAG, "jsonerror: " + e.getMessage());
			error.complete(e);
		}
		catch (IndexOutOfBoundsException e) {
		    android.util.Log.e(TAG, "index out of bounds");
		    error.complete(e);
		}
	}
	
	private boolean isLocationUnambiguous(String responseCode) {
        return responseCode.equals("100");
    }
	
	private boolean isBestGuessLocation(String responseCode) {
        return responseCode.equals("101");
    }

    private boolean isLargeLocation(String responseCode) {
        return responseCode.equals("110");
    }
    
    private boolean isLocationAmbiguous(String responseCode) {
        return responseCode.equals("200");
    }
    
    private boolean isLocationMisspelled(String responseCode) {
        return responseCode.equals("202");
    }
    
	private void fail(Callback<PropertyDataSourceResult> complete) {
		PropertyDataSourceResult result = new PropertyUnknownLocationResult();
		complete.complete(result);
	}
}

File: xamarinmvvmcross/PropertyCross.Core/Domain/Model/PropertyDataSource.cs
using System;
using Newtonsoft.Json.Linq;
using PropertyCross.Core.Domain.Services;

namespace PropertyCross.Core.Domain.Model
{
  public class PropertyDataSource
  {
    private IJsonPropertySearch _jsonPropertySearch;
      private readonly IMarshalInvokeService _marshalInvokeService;

      public PropertyDataSource(IJsonPropertySearch jsonPropertySearch, IMarshalInvokeService marshalInvokeService)
    {
        _jsonPropertySearch = jsonPropertySearch;
        _marshalInvokeService = marshalInvokeService;
    }

      public void FindProperties(double latitude, double longitude, int pageNumber, Action<PropertyDataSourceResult> callback, Action<Exception> error)
    {
      _jsonPropertySearch.FindProperties(latitude, longitude, pageNumber,
        response => HandleResponse(response, callback), error);
    }

    public void FindProperties(string searchText, int pageNumber, Action<PropertyDataSourceResult> callback, Action<Exception> error)
    {
      _jsonPropertySearch.FindProperties(searchText, pageNumber,
        response => HandleResponse(response, callback), error);
    }


    private void HandleResponse(string jsonResponse, Action<PropertyDataSourceResult> callback)
    {
      JObject json = JObject.Parse(jsonResponse);

      string responseCode = (string)json["response"]["application_response_code"];

      if (responseCode == "100" || /* one unambiguous location */
          responseCode == "101" || /* best guess location */
          responseCode == "110"  /* large location, 1000 matches max */)
      {
        var result = new PropertyListingsResult(json);
        _marshalInvokeService.Invoke(() => callback(result));
        

      }
      else if (responseCode == "200" || /* ambiguous location */
               responseCode == "202" /* mis-spelled location */)
      {
        var result = new PropertyLocationsResult(json);        
        _marshalInvokeService.Invoke(() => callback(result));

      }
      else
      {
        /*
        201 - unkown location
        210 - coordinate error
        */
          _marshalInvokeService.Invoke(() => callback(new PropertyUnknownLocationResult()));
      };
    }
  }
}
